### ISSUE TITLE
Rust Analyzer false positive: "expected pattern" in match expression 🐛

### ISSUE THREAD

**Issue by pcgeek86** [rust-lang/rust-analyzer#15487]
### Summary

The Rust Analyzer extension shows `Syntax Error: expected pattern`, when an unbounded [`RangeFrom` range pattern](https://doc.rust-lang.org/std/ops/struct.RangeFrom.html) is already present.

![image](https://github.com/rust-lang/rust-analyzer/assets/466713/67d7de3c-a400-48d4-bed7-1cc596bda054)

```rust
    let x = 5;
    let y: i32 = 0; 
    let z = 5;
    match x {
        z => {
            println!("It's five!");
        }
        6.. if y == 1 => {
        }
        6.. {
        }
    }

```

### Expected Behavior

Rust Analyzer should display **only** the error "_Expected: FAT_ARROW_".

### Actual Behavior

Rust Analyzer displays **two** errors:

- Expected: FAT_ARROW
- expected pattern

<img width="894" alt="image" src="https://github.com/rust-lang/rust-analyzer/assets/466713/0c787864-8893-49ed-a199-4343c3798a9e">

**rust-analyzer version**: `v0.3.1623`

**rustc version**: `rustc 1.70.0 (90c541806 2023-05-31)`

**relevant settings**: (eg. client settings, or environment variables like `CARGO`, `RUSTC`, `RUSTUP_HOME` or `CARGO_HOME`)

**Comment by feniljain** [comment]
Hello! 👋🏻 

I too had a look at this example:

<img width="475" alt="Screenshot 2023-08-21 at 09 18 12" src="https://github.com/rust-lang/rust-analyzer/assets/49019259/5ccb01af-bbed-421a-be0f-7959ef0c6c03">

and it seems to be correct, if you check in photo it does clearly mention, that it expects `=>`, `if` or `|` which is not just FAT_ARROW. Instead maybe second case of `..=4` should be checked as that only mentions `FAT_ARROW` case, either way I think diagnostics of any one case (preferrably second) need to be changed to match the other.

**Comment by pcgeek86** [comment]
Hello, thanks for taking a look at this! 👋🏻

> and it seems to be correct

I think that `6..` should not show `expected pattern`, because the pattern is already present. The other messages ("expected FAT_ARROW` and `expected one of ...` are correct though.

> either way I think diagnostics of any one case (preferrably second) need to be changed to match the other.

Yes, I agree. They should match behavior. Still, the `expected pattern` should not be displaying in either case, because the pattern is already present.

In general, only one message should be displayed. In this case, I think the most sensible message to display, in either case, is `expected one of ...`, because that demonstrates that there are multiple ways to solve the problem:

1. Add a fat arrow (simple)
2. Add an `|` or operator, to define multiple patterns
3. Add a match guard expression with `if`

**Comment by lnicola** [comment]
Related, we seem to accept `match x { 6..mut y => {} }`.
